spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc

```diff
@@ -164,6 +188,86 @@ Make sure to add 'spring-boot-configuration-processor' as an optional dependency
+----
+
+Note that the `spring-configuration-metadata.json` file is quite large. This is because it contains the concatenation of _all_ the properties that
+are available at runtime to the `log` sink (some of them come from `spring-boot-actuator.jar`, some of them come from
+`spring-boot-autoconfigure.jar`, even some more from `spring-cloud-starter-stream-sink-log.jar`, _etc._) Data Flow
+always relies on all those properties, even when a companion artifact is not available, but here all have been merged
+into a single file.
+
+To help with that (as a matter of fact, you don't want to try to craft this giant JSON file by hand), you can use the
+following plugin in your build:
+
+[source, xml]
+----
+<plugin>
+ 	<groupId>org.springframework.cloud</groupId>
+ 	<artifactId>spring-cloud-app-starter-metadata-maven-plugin</artifactId>
+ 	<executions>
+ 		<execution>
+ 			<id>aggregate-metadata</id>
+ 			<phase>compile</phase>
+ 			<goals>
+ 				<goal>aggregate-metadata</goal>
+ 			</goals>
+ 		</execution>
+ 	</executions>
+ </plugin>
+----
+
+NOTE: This plugin comes in _addition_ to the `spring-boot-configuration-processor` that creates the individual JSON files.
+Be sure to configure the two!
+
+The benefits of a companion artifact are manifold:
+
+1. being way lighter (usually a few kilobytes, as opposed to megabytes for the actual app), they are quicker to download,
+allowing quicker feedback when using _e.g._ `app info`
+2. as a consequence of the above, they can be used in resource constrained environments (such as PaaS) when metadata is
+the only piece of information needed
+3. finally, for environments that don't deal with boot uberjars directly (for example for Docker based runtimes), this
+is the only way to provide metadata about the properties supported by the app.
+
+Remember though, that this is entirely optional when dealing with uberjars. The uberjar itself is a source of metadata.
+
+==== Using the companion artifact
+Once you have a companion artifact at hand, you need to make the system aware of it so that it can be used.
+
+When registering a single app _via_ `app register`, you can use the optional `--metadata-uri` option in the shell, like so:
+
+[source]
+----
+dataflow:>app register --name log --type sink
+    --uri maven://org.springframework.cloud.stream.app:log-sink-kafka-10:1.2.0.RELEASE
+    --metadata-uri=maven://org.springframework.cloud.stream.app:log-sink-kafka-10:jar:metadata:1.2.0.BUILD-SNAPSHOT
+----
+
+When registering several files using the `app import` command, the file should contain a `<type>.<name>.metadata` line
+in addition to each `<type>.<name>` line. This is optional (_i.e._ if some apps have it but some others don't, that's fine).
+Here is an example for a Dockerized app:
+
+[source, properties]
+----
+...
+source.http=docker:springcloudstream/http-source-rabbit:latest
+source.http.metadata=maven://org.springframework.cloud.stream.app:http-source-rabbit:jar:metadata:1.2.0.BUILD-SNAPSHOT
```

Comment:
Since the companion artifact is a "maven artifact", could we consider elaborating that in the previous section? Someone who's all about Docker-based runtimes, they might see the "maven" reference ^^ and possibly get confused.